stan/math/rev/fun/gp_exp_quad_cov.hpp

```diff
@@ -208,82 +29,72 @@ class gp_exp_quad_cov_vari<T_x, double, T_l> : public vari {
  * @throw std::domain_error if sigma <= 0, l <= 0, or
  *   x is nan or infinite
  */
-template <typename T_x,
-          typename = require_arithmetic_t<typename scalar_type<T_x>::type>>
-inline Eigen::Matrix<var, -1, -1> gp_exp_quad_cov(const std::vector<T_x> &x,
-                                                  const var &sigma,
-                                                  const var &length_scale) {
-  check_positive("gp_exp_quad_cov", "sigma", sigma);
-  check_positive("gp_exp_quad_cov", "length_scale", length_scale);
-  size_t x_size = x.size();
-  for (size_t i = 0; i < x_size; ++i) {
+template <typename T_x, typename T_sigma, typename T_l,
+          require_all_stan_scalar_t<T_sigma, T_l>* = nullptr,
+          require_any_st_var<T_x, T_sigma, T_l>* = nullptr>
+inline Eigen::Matrix<var, -1, -1> gp_exp_quad_cov(const std::vector<T_x>& x,
+                                                  T_sigma sigma,
+                                                  T_l length_scale) {
+  check_positive("gp_exp_quad_cov", "marginal standard deviation", sigma);
+  check_positive("gp_exp_quad_cov", "length scale", length_scale);
+  for (size_t i = 0; i < x.size(); ++i) {
     check_not_nan("gp_exp_quad_cov", "x", x[i]);
   }
 
-  Eigen::Matrix<var, -1, -1> cov(x_size, x_size);
-  if (x_size == 0) {
-    return cov;
+  if (x.size() == 0) {
+    return Eigen::Matrix<var, -1, -1>();
   }
 
-  gp_exp_quad_cov_vari<T_x, var, var> *baseVari
-      = new gp_exp_quad_cov_vari<T_x, var, var>(x, sigma, length_scale);
+  double l_d = value_of(length_scale);
+  double sigma_d = value_of(sigma);
+  double sigma_sq_d = sigma_d * sigma_d;
+
+  arena_matrix<Eigen::MatrixXd> dist(x.size(), x.size());
+  arena_matrix<Eigen::MatrixXd> res_val(x.size(), x.size());
 
-  size_t pos = 0;
-  for (size_t j = 0; j < x_size - 1; ++j) {
-    for (size_t i = (j + 1); i < x_size; ++i) {
-      cov.coeffRef(i, j).vi_ = baseVari->cov_lower_[pos];
-      cov.coeffRef(j, i).vi_ = cov.coeffRef(i, j).vi_;
-      ++pos;
+  auto arena_x = to_arena_if<!is_constant<T_x>::value>(x);
```

Comment:
I now think this might be a bug. This is what you were saying in another comment right?

If this returns false, arena_x is simply x. arena_x gets used in the reverse pass callback (though maybe that branch of code doesn't execute) and therefore might get copied in.

Reply:
> I now think this might be a bug. This is what you were saying in another comment right?

Yeah, that is the same thing.